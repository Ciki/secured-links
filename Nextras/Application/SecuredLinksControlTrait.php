<?php

/**
 * This file is part of the Nextras community extensions of Nette Framework
 *
 * @license    MIT
 * @link       https://github.com/nextras
 * @author     Jan Skrasek
 */

namespace Nextras\Application;

use Nette;



trait SecuredLinksControlTrait
{

	/**
	 * {@inheritdoc}
	 */
	public function link($destination, $args = array())
	{
		if (!is_array($args)) {
			$args = func_get_args();
			array_shift($args);
		}

		list($destination, $args) = $this->getPresenter()->createSecuredRequest($this, $destination, $args);
		return parent::link($destination, $args);
	}



	/**
	 * For @secured annotated signal handler methods checks if URL parameters has not been changed
	 *
	 * @throws Nette\Application\UI\BadSignalException if there is not handler method or security token does not match
	 * @throws \RuntimeException if there is no redirect in a secured signal
	 */
	public function signalReceived($signal)
	{
		$method = $this->formatSignalMethod($signal);
		$reflection = new Nette\Reflection\Method($this, $method);
		if ($reflection->hasAnnotation('secured')) {
			$params = array();
			if ($this->params) {
				foreach ($reflection->getParameters() as $param) {
					if (isset($this->params[$param->name])) {
						$params[$param->name] = $this->params[$param->name];
					}
				}
			}
			if (!isset($this->params['_sec']) || $this->params['_sec'] !== $this->getPresenter()->getCsrfToken(get_class($this), $method, $params)) {
				throw new Nette\Application\UI\BadSignalException("Invalid security token for signal '$signal' in class {$this->reflection->name}.");
			}
		}

		parent::signalReceived($signal);

		if (isset($this->params['_sec'])) {
			throw new \RuntimeException("Secured signal '$signal' did not redirect. Possible csrf-token reveal by http referer header.");
		}
	}

}
