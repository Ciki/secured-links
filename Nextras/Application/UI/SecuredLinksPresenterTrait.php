<?php

/**
 * This file is part of the Nextras community extensions of Nette Framework
 *
 * @license    MIT
 * @link       https://github.com/nextras
 * @author     Jan Skrasek
 */

namespace Nextras\Application\UI;

use Nette;
use Nette\Application\UI\PresenterComponentReflection;



trait SecuredLinksPresenterTrait
{

	use SecuredLinksControlTrait;



	public function createSecuredRequest($component, $destination, $args)
	{
		// 1) fragment
		$a = strpos($destination, '#');
		if ($a === FALSE) {
			$fragment = '';
		} else {
			$fragment = substr($destination, $a);
			$destination = substr($destination, 0, $a);
		}

		// 2) ?query syntax
		$a = strpos($destination, '?');
		if ($a !== FALSE) {
			parse_str(substr($destination, $a + 1), $args); // requires disabled magic quotes
			$destination = substr($destination, 0, $a);
		}

		while (TRUE) {
			if ($component instanceof Nette\Application\UI\Presenter) {
				if (substr($destination, -1) !== '!') break;
			} else {
				if ($destination === 'this') break;
			}

			$signal = rtrim($destination, '!');
			$a = strrpos($signal, ':');
			if ($a !== FALSE) {
				$component = $component->getComponent(strtr(substr($signal, 0, $a), ':', '-'));
				$signal = (string) substr($signal, $a + 1);
			}
			if ($signal == NULL) { // intentionally ==
				throw new InvalidLinkException("Signal must be non-empty string.");
			}

			// only PresenterComponent - is it really needed?
			if (!$component instanceof Nette\Application\UI\PresenterComponent) break;

			$method = $component->formatSignalMethod($signal);

			$reflection = new PresenterComponentReflection(get_class($component));
			$signalReflection = $reflection->getMethod($method);

			if ($signalReflection->hasAnnotation('secured')) {
				$signalParams = array();
				if ($args) {
					// convert indexed parameters to named
					$_args = $args;
					self::argsToParams(get_class($component), $method, $_args);
					foreach ($signalReflection->getParameters() as $param) {
						if (isset($_args[$param->name])) {
							$signalParams[$param->name] = $_args[$param->name];
						}
					}
				}
				$args['_sec'] = $this->getCsrfToken(get_class($component), $method, $signalParams);
			}

			break;
		}

		return array($destination . $fragment, $args);
	}



	/**
	 * Returns unique token for method and params
	 * @param string
	 * @param string
	 * @param array
	 * @return string
	 */
	public function getCsrfToken($control, $method, $params)
	{
		$session = $this->getSession('Nette.Application.UI.Presenter/CSRF');
		if (!isset($session->token)) {
			$session->token = Nette\Utils\Strings::random();
		}

		$params = Nette\Utils\Arrays::flatten($params);
		$params = implode('|', array_keys($params)) . '|' . implode('|', array_values($params));
		return substr(md5($control . $method . $params . $session->token), 0, 8);
	}



	/**
	 * Converts list of arguments to named parameters.
	 * @param string class name
	 * @param string method name
	 * @param array arguments
	 * @param array supplemental arguments
	 * @return void
	 * @throws InvalidLinkException
	 */
	private static function argsToParams($class, $method, & $args, $supplemental = array())
	{
		$i = 0;
		$rm = new \ReflectionMethod($class, $method);
		foreach ($rm->getParameters() as $param) {
			$name = $param->getName();
			if (array_key_exists($i, $args)) {
				$args[$name] = $args[$i];
				unset($args[$i]);
				$i++;

			} elseif (array_key_exists($name, $args)) {
				// continue with process

			} elseif (array_key_exists($name, $supplemental)) {
				$args[$name] = $supplemental[$name];

			} else {
				continue;
			}

			if ($args[$name] === NULL) {
				continue;
			}

			$def = $param->isDefaultValueAvailable() && $param->isOptional() ? $param->getDefaultValue() : NULL; // see PHP bug #62988
			$type = $param->isArray() ? 'array' : gettype($def);
			if (!PresenterComponentReflection::convertType($args[$name], $type)) {
				throw new InvalidLinkException("Invalid value for parameter '$name' in method $class::$method(), expected " . ($type === 'NULL' ? 'scalar' : $type) . ".");
			}

			if ($args[$name] === $def || ($def === NULL && is_scalar($args[$name]) && (string) $args[$name] === '')) {
				$args[$name] = NULL; // value transmit is unnecessary
			}
		}

		if (array_key_exists($i, $args)) {
			$method = $rm->getName();
			throw new InvalidLinkException("Passed more parameters than method $class::$method() expects.");
		}
	}

}
