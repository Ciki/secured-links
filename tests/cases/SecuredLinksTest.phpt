<?php

use Nette\Application\Request;
use Nette\Application\Routers\SimpleRouter;
use Nette\Application\UI\Control;
use Nette\Application\UI\Presenter;
use Nette\Http\Request as HttpRequest;
use Nette\Http\Response;
use Nette\Http\Session;
use Nette\Http\UrlScript;
use Nextras\Application\UI\SecuredLinksControlTrait;
use Nextras\Application\UI\SecuredLinksPresenterTrait;
use Tester\Assert;

require __DIR__ . '/../bootstrap.php';


class TestControl extends Control
{
	use SecuredLinksControlTrait;
	/** @secured */
	public function handlePay($amount = 0)
	{
	}
}


class TestPresenter extends Presenter
{
	use SecuredLinksPresenterTrait;
	protected function startup()
	{
		parent::startup();
		$this['mycontrol'] = new TestControl;
	}
	public function renderDefault()
	{
		$this->terminate();
	}
	/** @secured */
	public function handlePay($amount = 0)
	{
	}
	/** @secured */
	public function handlePay2($amount)
	{
	}
	/** @secured */
	public function handleList(array $sections)
	{
	}
}


$url = new UrlScript('http://localhost/index.php');
$url->setScriptPath('/index.php');

$httpRequest = new HttpRequest($url);
$httpResponse = new Response();

$router = new SimpleRouter();
$request = new Request('Test', HttpRequest::GET, array());

$session = new Session($httpRequest, $httpResponse);
$section = $session->getSection('Nextras.Application.UI.SecuredLinksPresenterTrait');
$section->token = 'abcd';

$presenter = new TestPresenter();
$presenter->autoCanonicalize = FALSE;
$presenter->injectPrimary(NULL, NULL, $router, $httpRequest, $httpResponse, $session, NULL);
$presenter->run($request);


Assert::same( '/index.php?action=default&do=pay&presenter=Test&_sec=8607a814', $presenter->link('pay!') );
Assert::same( '/index.php?amount=200&action=default&do=pay&presenter=Test&_sec=8607a814', $presenter->link('pay!', [200]) );
Assert::same( '/index.php?amount=100&action=default&do=pay2&presenter=Test&_sec=948fe21d', $presenter->link('pay2!', [100]) );
Assert::same( '/index.php?amount=200&action=default&do=pay2&presenter=Test&_sec=03c6b49d', $presenter->link('pay2!', [200]) );
Assert::same( '/index.php?sections[0]=a&sections[1]=b&action=default&do=list&presenter=Test&_sec=9d4a84be', urldecode($presenter->link('list!', [['a', 'b']])) );
Assert::same( '/index.php?sections[0]=a&sections[1]=c&action=default&do=list&presenter=Test&_sec=fe7a715e', urldecode($presenter->link('list!', [['a', 'c']])) );

Assert::same( '/index.php?action=default&do=mycontrol-pay&presenter=Test&mycontrol-_sec=573011fa', $presenter['mycontrol']->link('pay') );
Assert::same( '/index.php?mycontrol-amount=200&action=default&do=mycontrol-pay&presenter=Test&mycontrol-_sec=573011fa', $presenter['mycontrol']->link('pay', [200]) );
