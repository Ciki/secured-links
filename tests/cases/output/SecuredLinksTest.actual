/index.php?amount=200&action=default&do=pay2&presenter=Test&_sec=03c6b49d