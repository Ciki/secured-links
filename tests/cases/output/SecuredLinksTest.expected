/index.php?amount=200&action=default&do=pay2&presenter=Test&_sec=8607a814